facade/src/main/java/org/jboss/pnc/facade/providers/BuildFetcher.java

```diff
@@ -0,0 +1,426 @@
+import org.jboss.pnc.spi.datastore.predicates.BuildRecordPredicates;
+import org.jboss.pnc.spi.datastore.repositories.BuildConfigSetRecordRepository;
+import org.jboss.pnc.spi.datastore.repositories.BuildConfigurationAuditedRepository;
+import org.jboss.pnc.spi.datastore.repositories.BuildConfigurationRepository;
+import org.jboss.pnc.spi.datastore.repositories.BuildRecordRepository;
+import org.jboss.pnc.spi.datastore.repositories.api.PageInfo;
+import org.jboss.pnc.spi.datastore.repositories.api.Predicate;
+import org.jboss.pnc.spi.datastore.repositories.api.SortInfo;
+import org.jboss.pnc.spi.datastore.repositories.api.impl.CursorPageInfo;
+import org.jboss.pnc.spi.exception.MissingDataException;
+import org.jboss.pnc.spi.exception.RemoteRequestException;
+
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static java.lang.Math.min;
+
+/**
+ * This class is used for fetching builds in an DB effective way. This means reading data in bulk and precisely
+ * computing what that bulk should contain and utilizes data preloading and caching.
+ */
+@RequestScoped
+@Slf4j
+public class BuildFetcher {
+
+    @Inject
+    private BuildMapper buildMapper;
+
+    @Inject
+    private BuildRecordRepository buildRecordRepository;
+
+    @Inject
+    private BuildCoordinator buildCoordinator;
+
+    @Inject
+    private BuildConfigSetRecordRepository buildConfigSetRecordRepository;
+
+    @Inject
+    private BuildConfigurationRepository buildConfigurationRepository;
+
+    @Inject
+    private BuildConfigurationAuditedRepository buildConfigurationAuditedRepository;
+
+    private Map<String, BuildTask> buildTaskCache = new HashMap<>();
+    private Map<Base32LongID, BuildRecord> buildRecordCache = new HashMap<>();
+    private Set<Base32LongID> dependenciesToFetch = new HashSet<>();
+
+    public BuildWithDependencies getBuildWithDependencies(String id) {
+        BuildTask buildTask = buildTaskCache.get(id);
+        if (buildTask != null) {
+            return new BuildWithDependencies(buildMapper.fromBuildTask(buildTask), buildTask);
+        }
+        Base32LongID dbid = buildMapper.getIdMapper().toEntity(id);
+        BuildRecord buildRecord = buildRecordCache.get(dbid);
+        if (buildRecord == null) {
+            buildRecord = buildRecordRepository.queryById(dbid);
+            if (buildRecord == null) {
+                throw new CorruptedDataException("Missing build with id:" + id);
+            }
+            log.warn("Prefetch missed build " + id);
```

Comment:
log.warn("Prefetch missed build {}", id);

Reply:
mine is faster, but ok